source/win32/low_latency_iocp_context.cpp

```diff
@@ -0,0 +1,682 @@
+                }
+            }
+
+            throw std::system_error{static_cast<int>(errorCode), std::system_category(), "GetQueuedCompletionStatusEx()"};
+        }
+
+        // Process completion-entries we received from the OS.
+        for (ULONG i = 0; i < completionEntriesRetrieved; ++i) {
+            auto& entry = completionBuffer[i];
+            if (entry.lpOverlapped != nullptr) {
+                win32::overlapped* o = reinterpret_cast<win32::overlapped*>(entry.lpOverlapped);
+                // TODO: Do we need to store the error-code/bytes-transferred here?
+
+                auto* state = to_io_state(o);
+
+                assert(state->pendingCompletionNotifications > 0);
+                if (--state->pendingCompletionNotifications == 0) {
+                    // This was the last pending notification for this state. 
+                    if (state->parent != nullptr) {
+                        // An operation is still attached to this state.
+                        // Notify it if we hadn't already done so.
+                        if (!state->completed) {
+                            state->completed = true;
+                            schedule_local(state->parent);
+                        }
+                    } else {
+                        // Otherwise the operation has already detached
+                        // from this I/O state and so we can immediately
+                        // return it to the pool.
+                        release_io_state(state);
+                    }
+                }
+            } else {
+                // A remote-thread wake-up notification.
+                shouldCheckRemoteQueue = true;
+            }
+        }
+    }
+}
+
+bool low_latency_iocp_context::try_dequeue_remote_work() noexcept {
+    auto items = remoteQueue_.dequeue_all_reversed();
+    if (items.empty()) {
+        return false;
+    }
+
+    do {
+        schedule_local(items.pop_front());
+    } while (!items.empty());
+
+    return true;
+}
+
+bool low_latency_iocp_context::poll_is_complete(vectored_io_state& state) noexcept {
+    // TODO: Double check the memory barriers/atomics we need to use
+    // here to ensure we perform this read, that it doesn't tear and
+    // that it has the appropriate memory semantics.
+
+    for (std::size_t i = 0; i < state.operationCount; ++i) {
+        overlapped* o = &state.operations[i];
+
+        // Mark volatile to indicate to the compiler that it might change in the
+        // background. The kernel might update it as the I/O completes.
+        volatile io_status_block* iosb = reinterpret_cast<io_status_block*>(o);
+        if (iosb->Status == STATUS_PENDING) {
+            return false;
+        }
+    }
+
+    // Make sure that we have visibility of the results of the I/O operations
+    // (assuming that the OS used a "release" memory semantic when writing
+    // to the 'Status' field).
+    std::atomic_thread_fence(std::memory_order_acquire);
```

Comment:
I suspect this is overkill, any syscall is an implicit sequentially consistent thread fence.

Reply:
I thought the idea was that when we poll we aren't making a syscall and so we need to do some synchronisation to ensure we have visibility of the (possibly other kernel thread's) writes to the buffer (at least for reads anyway). Maybe the barrier can be omitted from here?

An alternative I considered was just calling GetOverlappedResultEx() with a zero timeout and bAlertable=FALSE. Do you know if that ends up doing everything in user-space? Or were you manually reading the NTSTATUS because GetOverlappedResult() is too slow?

Reply:
I can't say if this particular atomic fence is necessary or not, I haven't traced through all the possible call paths. 

I don't call `GetOverlappedResultEx()` because it has a trivial implementation, see https://doxygen.reactos.org/dd/df8/iocompl_8c.html#ad5f55ae9b6f196ec2ae6cf775176fabe. And by not using it, I don't need to use `OVERLAPPED` which is larger than `IO_STATUS_BLOCK` because of the extra event handle, which you don't need as you can just loop waiting on the i/o handle and checking whether the status value has updated (I actually pump a single IOCP completion each loop too, if you're waiting anyway, might as well do something useful. But then mine is as-if blocking i/o, unlike yours).